Add documentation for aggregate functions

timabdulla opened the pull request:
Still WIP. Not ready for merge or review.

timabdulla commented on docs/references/api/aggregate_functions.rst:
```diff
@@ -0,0 +1,228 @@
+Using Columns from a Spread Resource
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When you :ref:`spread an embedded resource <spread_embed>`, the columns from the embedded resource that has been spread are treated as if they were columns of the top-level resource, both when using them as grouping columns and when applying aggregate functions to them.
```
@steve-chavez what is the proper terminology here for: "the embedded resource that has been spread"? would it be "the spreaded resource" or something else? finding it a bit awkward in this section

timabdulla replied:
thanks!

timabdulla commented on docs/references/configuration.rst:
```diff
@@ -178,6 +178,28 @@ app.settings.*
+
+  Proper safeguards could include:
+    - Use of a statement timeout. See :ref:`impersonated_settings`.
+    - Use of the `pg_plan_filter extension <https://github.com/pgexperts/pg_plan_filter>`_ to block excessively expensive queries.
```
should this link somewhere else or to the supabase fork? @steve-chavez

steve-chavez replied:
Looks good as it is :+1: 

(Supabase does have [pg_plan_filter](https://github.com/supabase/postgres#extensions) in their postgres image)

timabdulla commented:
@steve-chavez ready for some initial feedback if you get a chance, though I still have a bit more to finish. I also need to update the JSON formatting in the example response to match how it's done in the rest of the docs.

steve-chavez commented on docs/references/api/aggregate_functions.rst:
```diff
@@ -0,0 +1,228 @@
+
+When applying an aggregate function to a column, you are able to cast both the value of the input column *and* the value of the output of the aggregate function. In both cases, the syntax works as described in :ref:`casting_columns`, with the only difference being the placement of the cast.
+
+For instance, imagine that the ``amount`` column from the previous examples has the type ``text``. If you wanted to get the sum of this column, it would not be possible without first casting the value of the input column, as ``sum()`` does not support an input argument with the type ``text``. In this instance, you would need to cast the value of just the input column. Casting the the value of the ...
+
+.. tabs::
+
+  .. code-tab:: http
+
+    GET /orders?select=amount::numeric.sum() HTTP/1.1
```
Really useful considering we can't do a `sum()` over a json type.

steve-chavez commented on docs/references/api/aggregate_functions.rst:
```diff
@@ -0,0 +1,228 @@
+
+    curl "http://localhost:3000/orders?select=order_date,...customers(joined_date.max(),joined_date.min())
+
+As columns from a resource that has been spread are treated as if they were columns from the top-level resource, the ``max()`` and ``min()`` are applied *within* the context of the top-level, rather than within the context of the embedded resource, as in the previous section.
```
Shortening as per the discussion above
```suggestion
As columns from a spreaded resource are treated as if they were columns from the top-level resource, the ``max()`` and ``min()`` are applied *within* the context of the top-level, rather than within the context of the embedded resource, as in the previous section.
```

steve-chavez commented:
Btw, to pass the dictcheck, you can add `casted` to `postgrest.dict`.

timabdulla commented:
@steve-chavez Thanks for all the feedback. I think I've addressed all of your feedback, and I've made some further changes myself as well to improve clarity and readability, and I added the missing `TODO` items. I would now consider this to be complete, though it may need an additional read through in a few hours so I can review with fresh eyes.

One thing: There doesn't appear to be a consistent way to format JSON in the docs. There are a number of different styles in use. Not sure if there is a preference for how JSON should be formatted, but for now, I've just run all of my JSON through `jq` and used its pretty-printed output to standardize the formatting of my own examples.

timabdulla commented on docs/references/api/aggregate_functions.rst:
```diff
@@ -0,0 +1,341 @@
+The Case of ``count()``
+===========================
+
+.. note::
```
@steve-chavez I added this note. not sure if you agree with it, so let me know if you think it should be removed/modified.

steve-chavez replied:
Fully agree. Though there are lots of `count` users, not something we can break anytime soon. Still it's good to discourage its use.

steve-chavez reviewed (approved):
Great work!
